src/moonlink/src/storage/mooncake_table/batch_id_counter.rs

```diff
@@ -0,0 +1,178 @@
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+
+/// Batch ID counter for the two-counter allocation strategy.
+///
+/// The system uses two separate atomic counters to partition the 64-bit batch ID space:
+/// - **Streaming Counter**: Range 0 -> 2^63-1, used for streaming transactions
+/// - **Non-Streaming Counter**: Range 2^63+, used for regular operations
+///
+/// We give streaming batches the smaller range so that they are always behind the commit point, which points to the most recently added batch of the non-streaming batches.
+/// This ensures batch IDs are always monotonically increasing and unique across all transactions.
+pub(super) struct BatchIdCounter {
+    counter: Arc<AtomicU64>,
+    is_streaming: bool,
+}
+
+impl BatchIdCounter {
+    pub fn new(is_streaming: bool) -> Self {
+        Self {
+            counter: Arc::new(AtomicU64::new(if is_streaming { 0 } else { 1u64 << 63 })),
+            is_streaming,
+        }
+    }
+
+    pub fn load(&self) -> u64 {
```

Comment:
just for my own education, do you mind telling me why relaxed memory order is safe?